### PR TITLE
[TensorRT EP] Only instantiate TRT builder once 

### PR DESCRIPTION
**Pull request by chilo-ms**
The TRT builder instantization is slow (see [here](https://github.com/microsoft/onnxruntime/issues/18071)).
In current TRT EP, we instantiate builder object every time we need it. There are multiple places need the TRT builder so this causes huge performance overhead.

### CODE REVIEW

**Comment by chilo-ms** [comment]
Please see the github issue for more profiling details.

**Comment by gedoensmax** [comment]
Hm is Session creation supposed to be thread safe ? For TensorRT it would not be good practice to create multiple sessions in multiple threads on the same GPU. We could make the builder an unordered_map and store a builder per GPU id ? That would be a fair use case. 
Compiling 2 models at the same time on a single GPU is discouraged as it would mess with TensorRT's timing. 

Nonetheless do you think it would be a lot of work to make it thread safe - just to ensure safety. I assume it is not to simply protect it with a mutex ? We could return a tuple from the function with a scoped lock and the builder. `std::tie` afterwards for readability.
I believe the could be combined with the above approach of having a builder for each GPU, but might also be over engineering the problem.

**Comment by chilo-ms** [comment]
> Hm is Session creation supposed to be thread safe ? For TensorRT it would not be good practice to create multiple sessions in multiple threads on the same GPU. We could make the builder an unordered_map and store a builder per GPU id ? That would be a fair use case. Compiling 2 models at the same time on a single GPU is discouraged as it would mess with TensorRT's timing.

Session creation is not supposed to be thread-safe and i didn't see comments mentioned in ORT the C API saying it's thread-safe. 
yeah, creating mutiple sessions with multiple threads on the same GPU is not suggested, so we should skip this scenario. 
For multiple session for multiple GPUs, probably unordered_map to store a builder per GPU id is not needed since only one gpu device is maintained per session. 


> 
> Nonetheless do you think it would be a lot of work to make it thread safe - just to ensure safety. I assume it is not to simply protect it with a mutex ? We could return a tuple from the function with a scoped lock and the builder. `std::tie` afterwards for readability.

`nvinfer1::createInferBuilder` is not [mentioned ](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#threading)to be thread-safe. So, if we want GetBuilder() to be thread-safe, adding a mutex is needed. Moreover, considering if multiple threads want to call the same builder object's member function, for example `trt_builder.createNetworkV2()`, the calls seem need to be serialized (not parallel), or one builder object per thread seems needed.

Considering all the fair use cases, it seems to me that adding a mutex around `nvinfer1::createInferBuilder` should be enough.

**Comment by gedoensmax** [comment]
Yeah right forgot that multi GPU means multiple sessions.
Not sure I can follow here:
> if multiple threads want to call the same builder object's member function
Would the session internally use more than one thread at any point ? Looking at the code I just saw the member is not private - is there any reason for that ? 

Also I just thought about this sentence:
>  This PR makes TRT builder as a singleton.
For me a singleton would be something that is only instantiated once upon requesting it within a whole library. The singleton you have here is just a member variable. The only reason to not instantiate this in the Constructor is that there are ways in the future to possibly never need it so we only want to create it if really needed.

**Comment by chilo-ms** [comment]
> > if multiple threads want to call the same builder object's member function
> 
> Would the session internally use more than one thread at any point ? Looking at the code I just saw the member is not private - is there any reason for that ?

No, session won't internally use more than one thread. 
Just don't want to miss any cases so just asking about whether "multiple threads want to call the same builder object's member function" will happen. I also think this won't happen, so we can skip this case.
Good catch, i will make it private.



> For me a singleton would be something that is only instantiated once upon requesting it within a whole library. The singleton you have here is just a member variable. The only reason to not instantiate this in the Constructor is that there are ways in the future to possibly never need it so we only want to create it if really needed.

True. I removed it for not causing confusion.

**Comment by gedoensmax** [comment]
Sounds great then ! Not that my approval is needed, but I love it :D

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 1]
```diff
@@ -1209,6 +1209,17 @@ Status TensorrtExecutionProvider::OnRunEnd(bool sync_stream) {
   return Status::OK();
 }
 
+nvinfer1::IBuilder& TensorrtExecutionProvider::GetBuilder() const {
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1212]: should we document this isn't thread-safe and can only be called during single thread init phases ? (GetCapability() and Compile()) ?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 2]
```diff
@@ -1209,6 +1209,17 @@ Status TensorrtExecutionProvider::OnRunEnd(bool sync_stream) {
...
+  if (!builder_) {
+    TensorrtLogger& trt_logger = GetTensorrtLogger();
+    {
+      auto lock = GetApiLock();
+      builder_ = std::unique_ptr<nvinfer1::IBuilder>(nvinfer1::createInferBuilder(trt_logger));
+    }
+  }
+  return *builder_;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1220]: it's a little unusual to return reference?
Couldn't we return raw pointer?
Then many of the lines below wouldn't have to be changed. you could continue to use trt_builder->
also would be consistent with storing the raw pointer in the trt func_state

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc (hunk 1 above)
Comment (chilo-ms) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1212]: Yes, we should document this even though it doesn't make sense to have multiple threads running the inference session initialization.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**